ranger.py

```diff
@@ -149,3 +174,107 @@ def __init__(self, level):
             ],
             base_feats=base_feats,
         )
+
+class PrimalCompanion(Character):
+    def __init__(self, level: int, ranger: Character, attack: Weapon):
+        self.ranger = ranger
+        super().init(
+            level=level,
+            stats=[10, 10, 10, 10, 10, 10],
+            feats=[],
+            base_feats=[BeastChargeFeat(self), EquipWeapon(attack)],
+        )
+
+
+class BeastMaul(Weapon):
+    def __init__(self, base, **kwargs):
+        super().__init__(
+            name="Beast Maul", num_dice=1, die=8, mod="wis", base=base, **kwargs
+        )
+
+
+class BeastMasterAction(Feat):
+    def __init__(self, beast: Character, maul: Weapon, main_hand: Weapon, off_hand_nick: Weapon, off_hand_other: Weapon) -> None:
+        self.maul = maul
+        self.beast = beast
+        self.off_hand_other = off_hand_other
+        self.off_hand_nick = off_hand_nick
+        self.main_hand = main_hand
+        self.name = "BeastMasterAction"
+
+    def action(self, target: Target):
+        has_action = cast_spell(self.character, None)
+        if has_action:
+            if self.character.level >= 3:
+                if self.character.use_bonus("beast"):
+                    log.output(lambda: "Use bonus action for beast attack")
+                    self.beast.attack(target, self.maul)
+                    if self.character.level >= 11:
+                        self.beast.attack(target, self.maul)
+                    self.character.attack(target, self.main_hand)
+                else:
+                    if self.character.level >= 5:
+                        self.beast.attack(target, self.maul)
+                        if self.character.level >= 11:
+                            self.beast.attack(target, self.maul)
+                self.character.attack(target, self.main_hand)
+                self.character.attack(target, self.off_hand_nick)
+            else:
+                self.character.attack(target, self.main_hand)
+                if self.character.use_bonus("light weapon"):
+                    self.character.attack(target, self.off_hand_other)
```

Comment:
This should be tagged as `"light"` if the ranger doesn't have two weapon fighting at level 2. That will turn off the modifier damage